### ISSUE TITLE
TestFunctional/parallel/TunnelCmd: failed to hit nginx with DNS forwarded 

### ISSUE THREAD

**Issue by medyagh** [kubernetes/minikube#7809]
as seen here https://storage.googleapis.com/minikube-builds/logs/7780/e726215/HyperKit_macOS.html#fail_TestFunctional%2fparallel%2fTunnelCmd

```
fn_tunnel_cmd.go:141: failed to hit nginx with DNS forwarded "http://nginx-svc.default.svc.cluster.local": Temporary Error: Get http://nginx-svc.default.svc.cluster.local: dial tcp: lookup nginx-svc.default.svc.cluster.local on [2001:4860:4860::8844]:53: no such hos
```

also here

https://storage.googleapis.com/minikube-builds/logs/7780/e726215/VirtualBox_macOS.html#fail_TestFunctional%2fparallel%2fTunnelCmd

**Comment by tstromberg** [comment]
also, someone should remove this lie:

`fn_tunnel_cmd.go:148: tunnel at http://nginx-svc.default.svc.cluster.local is working!`

So, in any case, it seems like DNS forwarding may not be reliable currently, or at least immediately.

**Comment by govargo** [comment]
This issue still remains.

https://storage.googleapis.com/minikube-builds/logs/7843/ade5a44/HyperKit_macOS.html#fail_TestFunctional%2fparallel%2fTunnelCmd

```
helpers.go:234: (dbg) kubectl --context functional-20200422012827-4934 get po -A --show-labels:
-- stdout --
	NAMESPACE              NAME                                                     READY   STATUS              RESTARTS   AGE   LABELS
	default                busybox                                                  1/1     Running             0          20s   integration-test=busybox
	default                hello-node-677b9cfc6b-n6cvh                              0/1     ContainerCreating   0          33s   app=hello-node,pod-template-hash=677b9cfc6b
	default                mysql-78ff7d6cf9-bgtkb                                   1/1     Running             0          80s   app=mysql,pod-template-hash=78ff7d6cf9
	default                nginx-svc                                                1/1     Running             0          80s   run=nginx-svc
	kube-system            coredns-66bff467f8-6j5sx                                 1/1     Running             0          24m   k8s-app=kube-dns,pod-template-hash=66bff467f8
	kube-system            coredns-66bff467f8-wf6vb                                 1/1     Running             0          24m   k8s-app=kube-dns,pod-template-hash=66bff467f8
	kube-system            etcd-functional-20200422012827-4934                      1/1     Running             0          24m   component=etcd,tier=control-plane
	kube-system            kube-apiserver-functional-20200422012827-4934            1/1     Running             0          24m   component=kube-apiserver,tier=control-plane
	kube-system            kube-controller-manager-functional-20200422012827-4934   1/1     Running             0          24m   component=kube-controller-manager,tier=control-plane
	kube-system            kube-proxy-qqhzx                                         1/1     Running             0          24m   controller-revision-hash=c8bb659c5,k8s-app=kube-proxy,pod-template-generation=1
	kube-system            kube-scheduler-functional-20200422012827-4934            1/1     Running             0          24m   component=kube-scheduler,tier=control-plane
	kube-system            storage-provisioner                                      1/1     Running             0          24m   addonmanager.kubernetes.io/mode=Reconcile,integration-test=storage-provisioner
	kubernetes-dashboard   dashboard-metrics-scraper-84bfdf55ff-jz6g5               1/1     Running             0          6s    k8s-app=dashboard-metrics-scraper,pod-template-hash=84bfdf55ff
	kubernetes-dashboard   kubernetes-dashboard-bc446cc64-5c9gx                     1/1     Running             0          7s    k8s-app=kubernetes-dashboard,pod-template-hash=bc446cc64
```

Pod `nginx-svc` is running. And the error logs are not found.

I think this is DNS or Network related issue. If anyone have advice about more deep investigate, please let know me.

My idea:
 - check the service resource & endpoint resource of kubernetes.
 - check pod's /etc/resolv.conf

**Comment by govargo** [comment]
Is this related?

https://github.com/kubernetes/minikube/issues/7802

**Comment by govargo** [comment]
I tried `TestFunctional/parallel/TunnelCmd` test manually my minikube with docker driver on macOS.

As a result, `lookup nginx-svc.default.svc.cluster.local ... no such host` happened.

`minikube tunnel` on macOS changes `/etc/resolver/cluster.local`.
```
$ cat /etc/resolver/cluster.local
nameserver 10.96.0.10
search_order 1
```

`10.96.0.10` is kube-dns's cluster-ip  in kube-system.
```
$ kubectl get svc -n kube-system
NAME       TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   34m
```

But `kube-dns` service is type `ClusterIP`. So client on host cannot access `kube-dns`.
It's lead to `no such host` error.

**Comment by govargo** [comment]
My opinion to resolve this.

 1. make kube-dns acessible(like. ingress, nodePort, loadBalancer)

 2. add dns entry(like  "minikube ip": "xxxx.xxxxx.svc.cluster.local.") to `/etc/hosts`

**Comment by govargo** [comment]
Sorry, I tried above by docker driver on macOS.
I confirmed the dns resolve of `minikube tunnel` with hyperkit works correctly in my local pc.

**Comment by govargo** [comment]
After sending PR, the Hyperkit_macOS test failed by another reasion.

```
WithError(Unable to get machine status)=state: docker-machine-driver-hyperkit needs to run with elevated permissions. Please run the following command, then try again: sudo chown root:wheel /Users/jenkins/workspace/testdata/hyperkit-driver-without-version/docker-machine-driver-hyperkit && sudo chmod u+s /Users/jenkins/workspace/testdata/hyperkit-driver-without-version/docker-machine-driver-hyperkit
```
https://storage.googleapis.com/minikube-builds/logs/7956/0ee6042/HyperKit_macOS.html

-----

But the VirtualBox_macOS test reproduced this issue.

https://storage.googleapis.com/minikube-builds/logs/7956/0ee6042/VirtualBox_macOS.html

In this test, DNS resolution(`dig`) was success.

```
fn_tunnel_cmd.go:162: (dbg) Run:  dig +time=5 +tries=3 @10.96.0.10 nginx-svc.default.svc.cluster.local. A
fn_tunnel_cmd.go:172: DNS resolution for nginx-svc.default.svc.cluster.local. is working!
```

But `curl` access failed by `dial tcp: lookup nginx-svc.default.svc.cluster.local. on [2001:4860:4860::8844]:53: no such host`.
It looks like watching another DNS configuration.

```
fn_tunnel_cmd.go:180: failed to hit nginx with DNS forwarded "http://nginx-svc.default.svc.cluster.local.": Temporary Error: Get http://nginx-svc.default.svc.cluster.local.: dial tcp: lookup nginx-svc.default.svc.cluster.local. on [2001:4860:4860::8844]:53: no such host
```

After failing access, checked dns configuration looks no problem.
`/etc/resolver/cluster.local` is included successfully.

```
resolver #8
domain   : cluster.local
nameserver[0] : 10.96.0.10
flags    : Request A records, Request AAAA records
reach    : 0x00000002 (Reachable)
order    : 1
```

Still it's weird because the test still fails.

**Comment by medyagh** [comment]
seems to be closed by https://github.com/kubernetes/minikube/pull/7956

**Comment by govargo** [comment]
This is memo comment.

DNS resolution(dig) was success. but access for `nginx-svc.default.svc.cluster.local.` was failed by `no such host`.

```
fn_tunnel_cmd.go:162: (dbg) Run:  dig +time=5 +tries=3 @10.96.0.10 nginx-svc.default.svc.cluster.local. A
fn_tunnel_cmd.go:172: DNS resolution for nginx-svc.default.svc.cluster.local. is working!
```

So the dns configuration has no problem.
This issue's cause may be as a client side.
On local macOS, this issue doesn't happen. I think this was caused by test server's something.

E2E test uses net/http.
https://github.com/kubernetes/minikube/blob/93af9c1e43cab9618e301bc9fa720c63d5efa393/test/integration/fn_tunnel_cmd.go#L106-L121

And this http client queried to `[2001:4860:4860::8844]:53`. Off course that dns server doesn't know `nginx-svc.default.svc.cluster.local.`

```
fn_tunnel_cmd.go:180: failed to hit nginx with DNS forwarded "http://nginx-svc.default.svc.cluster.local.": Temporary Error: Get http://nginx-svc.default.svc.cluster.local.: dial tcp: lookup nginx-svc.default.svc.cluster.local. on [2001:4860:4860::8844]:53: no such host
```

I changed http client in order to use custom dns resolver. And now this works.

https://github.com/kubernetes/minikube/blob/f3d5c576e4d624c0c5bc7cee8499f65a5118a518/test/integration/fn_tunnel_cmd.go#L281-L298
